src/plugins/mavlink_ftp/include/plugins/mavlink_ftp/mavlink_ftp.h

```diff
@@ -0,0 +1,189 @@
+     * @param callback Callback to receive result of this request.
+     */
+    void download_async(std::string remote_file_path,
+                        std::string local_folder,
+                        progress_callback_t progress_callback,
+                        result_callback_t result_callback);
+
+    /**
+     * @brief Uploads local file to remote folder (asynchronous).
+     *
+     * @param local_file_path Local file to upload
+     * @param remote_folder Remote folder where uploaded file will be stored
+     * @param callback Callback to receive result of this request.
+     */
+    void upload_async(std::string local_file_path,
+                      std::string remote_folder,
+                      progress_callback_t progress_callback,
+                      result_callback_t result_callback);
+
+    /**
+     * @brief Downloads a vector of directory items from the system (asynchronous).
+     *
+     * @param path Path of a directory to list
+     * @param callback Callback to receive directory entries and result of this request.
+     */
+    void list_directory_async(std::string path, directory_items_and_result_callback_t callback);
+
+    /**
+     * @brief Creates directory (asynchronous).
+     *
+     * @param path Path of a directory to create
+     * @param callback Callback to receive result of this request.
+     */
+    void create_directory_async(std::string path, result_callback_t callback);
+
+    /**
+     * @brief Removes directory (asynchronous).
+     *
+     * @param path Path of a directory to remove
+     * @param callback Callback to receive result of this request.
+     */
+    void remove_directory_async(std::string path, result_callback_t callback);
+
+    /**
+     * @brief Removes file (asynchronous).
+     *
+     * @param path Path of a file to remove
+     * @param callback Callback to receive result of this request.
+     */
+    void remove_file_async(std::string path, result_callback_t callback);
+
+    /**
+     * @brief Truncates file (asynchronous).
+     *
+     * @param path Path of a file to remove
+     * @param offset Offset to which file is truncated
+     * @param callback Callback to receive result of this request.
+     */
+    void truncate_file_async(std::string path, uint32_t offset, result_callback_t callback);
+
+    /**
+     * @brief Renames fromPath to toPath (asynchronous).
+     *
+     * @param fromPath Path to rename
+     * @param toPath Destination path
+     * @param callback Callback to receive result of this request.
+     */
+    void rename_async(std::string from_path, std::string to_path, result_callback_t callback);
+
+    /**
+     * @brief Calculate CRC32 for file (asynchronous).
+     *
+     * @param path Path to file for which CRC32 is calculated.
+     * @param callback Callback to receive result of this request.
+     */
+    void calc_file_crc32_async(std::string path, file_crc32_result_callback_t callback);
```

Comment:
I don't think we need to expose the CRC check. The file transfer protocol should make sure everything is correct but the user of the library should not have to care in my opinion.

Reply:
I thought to use this function with camera definition xml file transfer so that we check if CRC has changed before we download it.